src/prof_recent.c

```diff
@@ -0,0 +1,407 @@
+ * that the associated recent allocation record can contain the following
+ * information:
+ * (1) The allocation is released;
+ * (2) The time of the deallocation; and
+ * (3) The prof_tctx associated with the deallocation.
+ */
+void
+prof_recent_alloc_reset(tsd_t *tsd, extent_t *extent) {
+	malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	prof_recent_t *recent = extent_prof_recent_alloc_get(extent);
+	malloc_mutex_unlock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+
+	if (recent == NULL) {
+		return;
+	}
+
+	prof_tctx_t *dalloc_tctx = prof_tctx_create(tsd);
+	if (dalloc_tctx != NULL) {
+		malloc_mutex_lock(tsd_tsdn(tsd), dalloc_tctx->tdata->lock);
+		++dalloc_tctx->recent_count;
+		dalloc_tctx->prepared = false;
+		malloc_mutex_unlock(tsd_tsdn(tsd), dalloc_tctx->tdata->lock);
+	}
+
+	malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	recent = extent_prof_recent_alloc_get(extent);
+	if (recent != NULL) {
+		extent_prof_recent_alloc_reset(extent, recent);
+		assert(nstime_compare(&recent->dalloc_time, &zero_time) == 0);
+		nstime_update(&recent->dalloc_time);
+		assert(recent->dalloc_tctx == NULL);
+		recent->dalloc_tctx = dalloc_tctx;
+	} else {
+		/* We lost the rase - the allocation record was just gone. */
+		decrement_recent_count(tsd, dalloc_tctx);
+	}
+	malloc_mutex_unlock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+}
+
+static void
+prof_recent_alloc_evict_extent(tsd_t *tsd, prof_recent_t *recent) {
+	malloc_mutex_assert_owner(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	if (recent->alloc_extent != NULL) {
+		extent_prof_recent_alloc_reset(recent->alloc_extent, recent);
+	}
+}
+
+static void
+assert_prof_recent_alloc_non_empty(tsd_t *tsd) {
+	malloc_mutex_assert_owner(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	assert(prof_recent_alloc_dummy != NULL);
+	assert(prof_recent_alloc_dummy->next != prof_recent_alloc_dummy);
+	assert(prof_recent_alloc_count > 0);
+}
+
+void
+prof_recent_alloc(tsd_t *tsd, prof_tctx_t *tctx, const void *ptr, size_t usize,
+    nstime_t t) {
+	extent_t *extent = iealloc(tsd_tsdn(tsd), ptr);
+	assert(extent != NULL);
+	assert(tctx != NULL && tctx == extent_prof_tctx_get(extent));
+
+	malloc_mutex_assert_not_owner(tsd_tsdn(tsd), tctx->tdata->lock);
+	malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+
+	if (prof_recent_alloc_max == 0) {
+		goto label_rollback;
+	}
+
+	if (prof_recent_alloc_dummy == NULL) {
+		/*
+		 * This is only encountered when the very first node is being
+		 * added - allocate the dummy node and circle it back.
+		 */
+		prof_recent_alloc_dummy = (prof_recent_t *)iallocztm(
+		    tsd_tsdn(tsd), sizeof(prof_recent_t),
+		    sz_size2index(sizeof(prof_recent_t)), false, NULL, true,
+		    arena_get(tsd_tsdn(tsd), 0, false), true);
+		if (prof_recent_alloc_dummy == NULL) {
+			goto label_rollback;
+		}
+		prof_recent_alloc_dummy->next = prof_recent_alloc_dummy;
+	}
+	assert(prof_recent_alloc_dummy != NULL);
+
+	{
+		/* Fill content into the dummy node. */
```

Comment:
Will it be better this way: fill the actual new node; and keep the dummy node statically defined (either initialize statically or in prof_boot somewhere.) and never changes its location; the only thing changes is where the dummy nodes points to in the actual circular list. (then maybe call the dummy node as head instead).